### ISSUE TITLE
Importing Data with Date Format imports the data as Null

### ISSUE THREAD

**Issue by kapee96** [mbleron/ExcelTable#45]
While importing the Data from XSLX files the data where the cell contains a Date is getting imported as NULL

E.x. of the data columns in XLSX

Product Name	        Lot Details		Date Of import	Date of Packing
Face Masks		BoxABC			16-Jul-2015		n/a
Face Masks		BoxXYZ			16-Jul-2016		n/a
Face Masks		Box123			16-Jul-2017		n/a
Face Masks		Box456			16-Jul-2018		n/a

Code for Import

INSERT INTO TDL_CLIENT_DATA (PRODUCT_NAME, LOT_NAME, DATE_IMPORT, PACK_DATE)
SELECT t.* 
FROM Table(
       ExcelTable.getRows(ExcelTable.getFile('DMP_DIR','Face_Masks.xlsx')
   , 'Lot Details', 
   '"COL1" VARCHAR2(2000),
   "COL2" VARCHAR2(2000),
   "COL3" VARCHAR2(2000),
   "COL4" VARCHAR2(2000)'
      )
   ) t
;

if i change the column to date format then i get the error

INSERT INTO TDL_CLIENT_DATA (PRODUCT_NAME, LOT_NAME, DATE_IMPORT, PACK_DATE)
SELECT t.* 
FROM Table(
       ExcelTable.getRows(ExcelTable.getFile('DMP_DIR','Face_Masks.xlsx')
   , 'Lot Details', 
   '"COL1" VARCHAR2(2000),
   "COL2" VARCHAR2(2000),
   "COL3" Date format 'DD-MMM-YYYY',
   "COL4" Date format 'DD-MMM-YYYY''
      )
   ) t
;

ORA-00907: missing right parenthesis
00907. 00000 -  "missing right parenthesis"
*Cause:    
*Action:

Any help would be appreciated

**Comment by mbleron** [comment]
Could you please attach a sample file to reproduce the issue?

As for the error in the 2nd statement, that's because the single quotes aren't escaped around the date formats.

**Comment by kapee96** [comment]
Attached is the same
[Lot Details.xlsx](https://github.com/mbleron/ExcelTable/files/12552480/Lot.Details.xlsx)

For the date issue on the Single quotes how do i bypass since it was the code provided in your sample code format

**Comment by mbleron** [comment]
> For the date issue on the Single quotes how do i bypass since it was the code provided in your sample code format

Quotes within a string literal can be escaped by doubling them : 

```
'"COL1" VARCHAR2(2000),
"COL2" VARCHAR2(2000),
"COL3" Date format ''DD-MMM-YYYY'',
"COL4" Date format ''DD-MMM-YYYY'''
```

or by using an alternate quoting mechanism, e.g.
```
q'{
"COL1" VARCHAR2(2000),
"COL2" VARCHAR2(2000),
"COL3" Date format 'DD-MMM-YYYY',
"COL4" Date format 'DD-MMM-YYYY'
}'
```

there are examples of that in the README too

**Comment by mbleron** [comment]
The problem is you have a mix of dates and strings in the same column. 
ExcelTable won't handle that situation correctly, and SQL does not anyway, that's the difference between a spreadsheet and a relational data set. In the latter, the data type must be the same for a given column, unless the data is encapsulated in a "variant" type such as ANYDATA.

If you project as VARCHAR2, ExcelTable doesn't convert date to string because there's no direct match between Excel date format and Oracle's. I could probably workaround that by converting to string using the session NLS_DATE_FORMAT.

On the other hand, if you project as DATE, it'll error out since there's non-date data in the column ('n/a' for instance).
Again, I could add an option to catch the conversion error and return NULL instead.

Last, I think the best way is to introduce ANYDATA projection so that the original data is preserved as much as possible.

I'll keep this issue open as an enhancement to implement above ideas.

**Comment by lee-lindley** [comment]
I looked back at work I did on this problem. I found a way, but it was terribly messy.  If you can refine the concepts to make ExcelTable and ExcelGen work together for this hard problem, I'm sure I won't be the only one who is grateful.

 

https://lee-lindley.github.io/oracle/plsql/2022/09/18/spreadsheets_with_plsql.html

 

-- 

Lee Lindley

***@***.***

**Comment by kapee96** [comment]
Thank you for looking into this. I receive data from various partner across different countries and all have the date format differently hence my goal was to just load it as is and then do the correction in Oracle rather than updating these in individual XLS file.

**Comment by mbleron** [comment]
The date format itself is not important because Excel do not store dates as strings but as numbers.

In the upcoming version, ExcelTable will be able to ignore invalid date values in a resulting SQL DATE column.

**Comment by mbleron** [comment]
A bit of bitterness here.
It seems that ANYDATA projection doesn't work.
On 12.1, Oracle doesn't recognize the SYS.ANYDATA built-in type while creating the transient object type supporting our projection :
```
  "ID"               number(2)
, "STRING_VALUE"     varchar2(2000)
, "DATE_VALUE"       anydata
, "TIMESTAMP_VALUE"  timestamp  NULL ON ERROR
, "NUMBER_VALUE"     number  NULL ON ERROR
```
_behind the scenes we get an "UNKNOWN TYPE" :_
```
PARSING IN CURSOR #140253763008328 len=189 dep=1 uid=0 oct=77 lid=0 tim=688952424141 hv=1040524152 ad='7f8f5fb14c28' sqlid='8nf1pb0z0a8vs'
CREATE TYPE "DEV"."SYSTPBOsSIPwJbGXgY2VeqMCrBw==" AS OBJECT ("ID" NUMBER(2), "STRING_VALUE" VARCHAR2(2000), "DATE_VALUE" UNKNOWN TYPE, "TIMESTAMP_VALUE" TIMESTAMP(6), "NUMBER_VALUE" NUMBER)
END OF STMT
```

Encapsulating the ANYDATA value in a user-defined object type works, but that's for 12.1 only.
On 19.3, we get this exotic, never-seen-before, error : 
`ORA-32766: instr with negative offset: use varchar semantics on LOBs`
raised by `ANYTYPE.GetPersistent` method, which has apparently ceased to function altogether.

~~Very sad.~~
Ok, should have RTFM before... the method has been replaced by [GETANYTYPEFROMPERSISTENT](https://docs.oracle.com/en/database/oracle/oracle-database/19/arpls/ANYTYPE-TYPE.html#GUID-B921DE72-78B7-4EF5-88C0-8E955B04F6D1) function.  
So there's still hope 😃

**Comment by lee-lindley** [comment]
The anytype implementation is less "helpful" than most tools Oracle gives us. It is also implemented only in the SQL engine AFAIK, so calls to the methods from PL/SQL are slow. Working with anytype is painful. Your ExcelTable implementation is the most advanced user of anyType that I've run across. Now I don't feel so lightweight for punting on it last year.

**Comment by mbleron** [comment]
I've been adding some new [features](https://github.com/mbleron/ExcelTable#columns-syntax-specification) to help dealing with this issue.

I'll try to sum it up using this little sample data ([test_on_error.xlsx](https://github.com/mbleron/ExcelTable/blob/master/samples/test_on_error.xlsx)):

![test-on-error](https://github.com/mbleron/ExcelTable/assets/19926388/b83dc4d8-7e25-45d1-bd4c-ff1ba3c15f9e)

* Column C holds date values except for row 6 and 8 which have text values.
* Column D holds timestamp values except for row 10.
* Column E holds number values expect for row 5.

By default, ExcelTable raises an error if it finds a value that cannot be converted to the requested data type, for instance : 
```
SELECT t.*
FROM Table(
       ExcelTable.getRows(
         ExcelTable.getFile('TEST_DIR','test_on_error.xlsx')
       , 'data'
       , q'{
           "ID"               number(2)
         , "STRING_VALUE"     varchar2(2000)
         , "DATE_VALUE"       date
         , "TIMESTAMP_VALUE"  timestamp
         , "NUMBER_VALUE"     number
         }'
       , 'A2'
       )
     ) t
;

ORA-20722: Invalid NUMBER value in cell E5 : yyy
```
Now comes the new NULL ON ERROR clause:
```
SELECT t.*
FROM Table(
       ExcelTable.getRows(
         ExcelTable.getFile('TEST_DIR','test_on_error.xlsx')
       , 'data'
       , q'{
           "ID"               number(2)
         , "STRING_VALUE"     varchar2(2000)
         , "DATE_VALUE"       date       null on error
         , "TIMESTAMP_VALUE"  timestamp  null on error
         , "NUMBER_VALUE"     number     null on error
         }'
       , 'A2'
       )
     ) t
;
 
 ID STRING_VALUE   DATE_VALUE  TIMESTAMP_VALUE                NUMBER_VALUE
--- -------------- ----------- ------------------------------ ------------
  1 ]G[m*t\p(y     08/09/2023  08-SEP-23 11.27.30.123000 AM   7.4959172902
  2 cXhRjUsa:p                 08-SEP-23 12.27.30.123000 PM   4.0812586139
  3 (~~47-x66v     10/09/2023  08-SEP-23 01.27.30.123000 PM   2.9020782461
  4 Q}<'.(a6->     11/09/2023  08-SEP-23 02.27.30.123000 PM   
  5 u({"[kKHzJ                 08-SEP-23 03.27.30.123000 PM   7.2229577149
  6 E?ajVtTU%A     13/09/2023  08-SEP-23 04.27.30.123000 PM   3.4215571945
  7 l.nD"r`Kiw                 08-SEP-23 05.27.30.123000 PM   1.5030854157
  8 sji6rdkK]2     15/09/2023                                 5.3662811056
  9 X?Vxb)sq]3     16/09/2023                                 
 10 JOE|cxz]MF     17/09/2023  08-SEP-23 08.27.30.123000 PM   2.9953452386
 
10 rows selected.
```
Inconvertible values got nullified.
Note that if some values can be converted via an explicit mask (e.g. cell C8), we can still add it in the format clause : 
`"DATE_VALUE"       date   format 'YYYY-MM-DD'    null on error`

The other way to handle heterogeneous data types is to encapsulate them in an object instance.
The straightforward approach would be to allow for ANYDATA type projection, but unfortunately, due to some internal restrictions - or dare I say "bug" - it's not possible. 
So I created a new user-defined object type called ExcelVariant, that is accepted by the ODCI framework. Basically it's just a container for an ANYDATA value, with some convenience getters for the supported SQL types.
Here's a sample usage which produces the same result as the last query:
```
SELECT t.id
     , t.string_value
     , t.date_value.getDate(nullOnError => 1) as date_value
     , t.timestamp_value
     , t.number_value
FROM Table(
       ExcelTable.getRows(
         ExcelTable.getFile('TEST_DIR','test_on_error.xlsx')
       , 'data'
       , q'{
           "ID"               number(2)
         , "STRING_VALUE"     varchar2(2000)
         , "DATE_VALUE"       VARIANT
         , "TIMESTAMP_VALUE"  timestamp  NULL ON ERROR
         , "NUMBER_VALUE"     number     NULL ON ERROR
         }'
       , 'A2'
       )
     ) t
;
```

Hope that helps.

**Comment by mbleron** [comment]
Closing this one. Features implemented in v5.4.1.
